src/tactic/interactive_expr.lean

```diff
@@ -65,6 +67,80 @@ meta def sf.of_eformat : eformat → sf
+| [] _ := none   -- (h::t) ++ _ ≠ []
+-- if t₂ ++ z = t₁ then (h₁ :: t₁) ++ z = (h₁ :: t₂)
+| (h₁ :: t₁) (h₂ :: t₂) := if h₁ = h₂ then subtract t₁ t₂ else none
+
+/-- Get the subexpression and sf object at the given address, if any.
+The address must point to a `sf.tag_expr` boundary. -/
+meta def sf.follow : expr.address → sf → option (expr × sf)
+| [] s := none
+| l (sf.tag_expr ea e  m) := do
+  a ← subtract l ea,
+  if a = [] then pure (e,m) else
+  sf.follow a m
+| l (sf.block _ a) := sf.follow l a
+| l (sf.highlight _ a) := sf.follow l a
+| l (sf.of_string _) := none
+| l (sf.compose a b) := sf.follow l a <|> sf.follow l b
+
+/-- Run `f` on each of the immediate child substrings of the given string. -/
+meta def sf.traverse {m} [applicative m] (f : sf → m sf) : sf → m sf
+| (sf.tag_expr ea e m) := pure (sf.tag_expr ea e) <*> f m
+| (sf.block x a) := pure (sf.block x) <*> f a
+| (sf.highlight x a) := pure (sf.highlight x) <*> f a
+| (sf.compose a b) := pure (sf.compose) <*> f a <*> f b
+| (sf.of_string s) := pure $ sf.of_string s
+
+/-- Run `f` on each child. If it fails then stop traversing and keep it the same.-/
+meta def sf.replace {m} [monad m] [alternative m] (f : sf → m sf) : sf → m sf
+| x := (f x >>= sf.traverse sf.replace) <|> pure x
+
+/-- The test for whether the proposition is a valid target for restricted quantifier collapsing.
+Eg (@has_mem.mem _ _ _ _) -/
+meta def sf.collapse_restricted_quantifiers_pred : expr → tactic bool
+| `(@gt _ _ %%(expr.var 0) _) := pure tt
+| `(@has_lt.lt _ _ %%(expr.var 0) _) := pure tt
+| `(@has_le.le _ _ %%(expr.var 0) _) := pure tt
+| `(@ge _ _ %%(expr.var 0) _) := pure tt
+| `(@has_mem.mem _ _ _ %%(expr.var 0) _) := pure tt
+-- [fixme] add to this list! So many predicates. Maybe just include all of them.
+| _ := pure ff
+
+open expr.coord
+
+/-- If the given sf has the form `∀ (x : α), P x → Q`
+and `P x` obeys `collapse_restricted_quantifiers_pred`
+then replace it with `∀ (P x), Q`, mapping the addresses properly
+so interactive rendering still works. -/
+meta def sf.collapse_restricted_quantifiers_core : sf → tactic sf
+| s@(sf.tag_expr addr e@(expr.pi x xbi xy (expr.pi p pbi py b)) a) := (do
+  should_collapse ← sf.collapse_restricted_quantifiers_pred py,
+  if ¬ should_collapse then pure s else do
+  a1 ← pure $ [pi_body, pi_var_type],
+  a2 ← pure $ [pi_body, pi_body],
+  (e1, s1) ← sf.follow a1 a,
+  (e2, s2) ← sf.follow a2 a,
+  pure (sf.tag_expr addr e $ "∀ (" ++ (sf.tag_expr (addr ++ a1) e1 s1) ++ "), " ++ (sf.tag_expr (addr ++ a2) e2 s2))
+  ) <|> pure s
+| s@(sf.tag_expr addr e a) := (do
+  `(Exists %%pred) ← pure e,
+  expr.lam _ _ α pred ← pure pred,
+  `(Exists %%pred) ← pure pred,
+  expr.lam _ _ py pred ← pure pred,
+  should_collapse ← sf.collapse_restricted_quantifiers_pred py,
+  if ¬ should_collapse then pure s else do
+  a1 ← pure $ [app_arg, lam_var_type], -- [hack] There is a bug in the addresing code. Second binder has the address of the first binder!
```

Comment:
Is this bug reported somewhere more visible?

Reply:
Fixed now in https://github.com/leanprover-community/lean/pull/520

Reply:
This code will need to be changed once the fix has made it downstream

Reply:
I've marked this dependent on #5895, which migrates mathlib to 3.26.0c which includes the fix.